File: funcs.jl
using Random
using Statistics
using StatsBase

include("constants.jl")

# Generate inter-arrival times for each center using an exponential process,
# then return all patients sorted by arrival time.
# Returns an N×3 matrix: [patient_id, arrival_time, center_id]
function generate_patient_arrivals(
    rates::Vector{Float64},
    center_activations::AbstractVector{<:Integer},
    num_centers::Int,
    num_patients::Int
)
    all_arrivals = Matrix{Float64}(undef, 0, 3)
    for i in 1:num_centers
        # Inverse-transform sampling: u ~ Uniform(0,1) → Exponential inter-arrivals
        u = rand(num_patients)
        inter_arrivals = -log.(1 .- u) ./ rates[i]
        arrival_times  = cumsum(inter_arrivals) .+ center_activations[i] * 30
        block = hcat(
            Float64.((i-1)*num_patients .+ (1:num_patients)),
            arrival_times,
            fill(Float64(i), num_patients)
        )
        all_arrivals = vcat(all_arrivals, block)
    end
    return all_arrivals[sortperm(all_arrivals[:, 2]), :]
end

# Assign each patient to a stratum: stratum 1 with probability p, stratum 2 otherwise.
function generate_strata_assignments(num_patients::Int, p::Float64)
    return [u ≤ p ? Int8(1) : Int8(2) for u in rand(num_patients)]
end

# Generate a permuted block design treatment sequence.
# Returns a 1×N matrix. For stratified use, the caller reshapes and transposes
# to get a (num_strata × positions_per_stratum) matrix.
function generate_treatment_blocks(
    ratio::Tuple,
    sample_size::Int,
    treat_arms::Int,
    block_size::Int
)
    block_template = [t for t in 1:treat_arms for _ in 1:(ratio[t] * block_size ÷ sum(ratio))]
    n_blocks = ceil(Int, 2 * sample_size / block_size)
    return reduce(hcat, [reshape(shuffle(block_template), 1, :) for _ in 1:n_blocks])
end

# Return supply and FR parameters for a given scenario (1–12).
#
# Scenario layout:
#   Rows (IRT spec):     F0a (1–3), F0b (4–6), F1a (7–9), F1b (10–12)
#   Columns (supply):    Low, Medium, High
#
# Returns: (resupply_amount, init_supply, critical_point, fr_allowed, backfill_enabled, cap)
function scenario_params(scenario::Int, cap::Int)
    supply_configs = [
        (LOW_RESUPPLY,  copy(LOW_INIT),  LOW_CRITICAL),
        (MED_RESUPPLY,  copy(MED_INIT),  MED_CRITICAL),
        (HIGH_RESUPPLY, copy(HIGH_INIT), HIGH_CRITICAL),
    ]
    fr_configs = [
        (false, false, 0),    # F0a: no forced randomization; send home if supply missing
        (true,  false, 0),    # F0b: FR enabled but cap = 0 (never triggers)
        (true,  false, cap),  # F1a: FR enabled, skip forward without backfilling
        (true,  true,  cap),  # F1b: FR enabled, skip forward with backfilling
    ]
    supply_idx = ((scenario - 1) % 3) + 1
    fr_idx     = ((scenario - 1) ÷ 3) + 1
    resupply, init_supply, crit = supply_configs[supply_idx]
    fr_allowed, backfill, eff_cap = fr_configs[fr_idx]
    return resupply, init_supply, crit, fr_allowed, backfill, eff_cap
end

# 90th quantile approximation: mean + 1.645 * std, computed along a given dimension.
function quantile_90(values; dims::Int)
    return mean(values, dims=dims) .+ 1.645 .* std(values, dims=dims)
end

File: rshiny/R/funcs.R
# Generate patient arrivals via inverse-transform exponential sampling.
# Returns an N×3 matrix: [patient_id, arrival_time, center_id]
generate_patient_arrivals <- function(rates, center_activations, num_centers, num_patients) {
  blocks <- vector("list", num_centers)
  for (i in seq_len(num_centers)) {
    u              <- runif(num_patients)
    inter_arrivals <- -log(1 - u) / rates[i]
    arrival_times  <- cumsum(inter_arrivals) + center_activations[i] * 30
    blocks[[i]]    <- cbind(
      (i - 1L) * num_patients + seq_len(num_patients),
      arrival_times,
      rep(i, num_patients)
    )
  }
  all_arrivals <- do.call(rbind, blocks)
  all_arrivals[order(all_arrivals[, 2L]), ]
}

# Assign each patient to a stratum. Returns an integer vector of stratum indices.
# Matches Julia's findfirst(>=(u), cumprobs) logic.
generate_strata_assignments <- function(num_patients, probs) {
  cum_probs <- cumsum(probs)
  u         <- runif(num_patients)
  # rowSums counts how many cumprobs u exceeds; +1 gives 1-based stratum index
  as.integer(rowSums(outer(u, cum_probs, ">")) + 1L)
}

# Generate a permuted block treatment sequence.
# Returns a flat integer vector of length num_strata * sample_size.
# Caller reshapes: t(matrix(raw, nrow = sample_size, ncol = num_strata))
generate_treatment_blocks <- function(ratio, sample_size, treat_arms, block_size, num_strata) {
  block_template <- rep(seq_len(treat_arms), times = ratio * (block_size %/% sum(ratio)))
  n_blocks       <- ceiling(num_strata * sample_size / block_size)
  unlist(replicate(n_blocks, sample(block_template), simplify = FALSE))
}

# Return supply and FR parameters for a given scenario number (1-12).
scenario_params <- function(scenario, params) {
  supply_configs <- list(
    list(resupply_amt = params$low_resupply,  init_supply = params$low_init,  critical_pt = params$low_critical),
    list(resupply_amt = params$med_resupply,  init_supply = params$med_init,  critical_pt = params$med_critical),
    list(resupply_amt = params$high_resupply, init_supply = params$high_init, critical_pt = params$high_critical)
  )
  fr_configs <- list(
    list(fr_allowed = FALSE, backfill_enabled = FALSE, cap = 0L),
    list(fr_allowed = TRUE,  backfill_enabled = FALSE, cap = 0L),
    list(fr_allowed = TRUE,  backfill_enabled = FALSE, cap = params$initial_cap),
    list(fr_allowed = TRUE,  backfill_enabled = TRUE,  cap = params$initial_cap)
  )
  supply_idx <- (scenario - 1L) %% 3L + 1L
  fr_idx     <- (scenario - 1L) %/% 3L + 1L
  c(supply_configs[[supply_idx]], fr_configs[[fr_idx]])
}

# 90th quantile approximation: mean + 1.645 * sd, over the first dimension of an array.
quantile_90 <- function(x, dims) {
  apply(x, dims, mean) + 1.645 * apply(x, dims, sd)
}
